python/cudf/cudf/_libxx/join.pyx

```diff
@@ -0,0 +1,90 @@
+import pandas as pd
+
+from cudf._libxx.lib cimport *
+from cudf._libxx.table cimport Table
+from cudf._libxx.join cimport *
+
+from cudf._lib.utils cimport *
+from cudf._lib.utils import *
+
+
+cpdef join(lhs, rhs, left_on, right_on, how, method):
+    """
+      Call libcudf++ join for full outer, inner and left joins.
+      Returns a list of tuples [(column, valid, name), ...]
+    """
+    if how not in ['left', 'inner', 'outer']:
+        msg = "new join api only supports left, inner or outer"
+        raise ValueError(msg)
+
+    left_idx = []
+    right_idx = []
+
+    assert(len(left_on) == len(right_on))
+
+    cdef vector[int] left_on_ind
+    cdef vector[int] right_on_ind
+    cdef vector[pair[int, int]] columns_in_common
+
+    result_col_names = []  # Preserve the order of the column names
+
+    for name, col in lhs._data.items():
+        check_gdf_compatibility(col)
+        result_col_names.append(name)
+
+    for name in left_on:
+        # This will ensure that the column name is valid
+        lhs._data[name]
+        left_on_ind.push_back(list(lhs._data.keys()).index(name))
+        if (name in right_on and
+           (left_on.index(name) == right_on.index(name))):
+            columns_in_common.push_back(pair[int, int](
+                list(lhs._data.keys()).index(name),
+                list(rhs._data.keys()).index(name)))
+
+    for name in right_on:
+        # This will ensure that the column name is valid
+        rhs._data[name]
```

Comment:
Add a more descriptive assert here too

Reply:
@trevorsm7 should be fixed now - we'll raise up front with something a bit more explicit.